### ISSUE TITLE
'ETIMEDOUT',   errno: -4039,   code: 'ETIMEDOUT',   syscall: 'connect',   address: '20.205.243.166',   port: 443,   caller: 'git.clone'

### ISSUE THREAD

**Issue by jahnli** [isomorphic-git/isomorphic-git#1487]
errno: -4039,
  code: 'ETIMEDOUT',
  syscall: 'connect',
  address: '20.205.243.166',
  port: 443,
  caller: 'git.clone'

I've set up the Git proxy, but it still doesn't work

http.https://github.com.proxy=socks5://127.0.0.1:7890
http.proxy=http://127.0.0.1:7890
https.proxy=https://127.0.0.1:7890

But  isomorphic-git    Git config Proxy is not executed

**Comment by mojavelinux** [comment]
isomorphic-git does not have direct support for using a proxy. You need to provide a http plugin that integrates proxy support. You can see how I did it in Antora here: https://gitlab.com/antora/antora/-/blob/main/packages/content-aggregator/lib/git-plugin-http.js

**Comment by jahnli** [comment]
provide a http plugin that integrates proxy support. 

superagent-proxy   or  https-proxy-agent     ????

**Comment by mojavelinux** [comment]
I prefer hpagent.

**Comment by jahnli** [comment]
Is it possible to replace const HTTP = require("isomorphic-git/ HTTP /node") with http-proxy-agent?

**Comment by mojavelinux** [comment]
No, the http plugin requires a special signature. But it can be a very simple wrapper that injects the agent into the argument list.

**Comment by jahnli** [comment]
Is it possible to provide a tutorial or demo in the official document

**Comment by mojavelinux** [comment]
I have offered what I have time to provide, which is a working implementation of an http plugin that you can use with isomorphic-git.

**Comment by jcubic** [comment]
@jahnli you don't need a tutorial just look at the code: https://github.com/isomorphic-git/isomorphic-git/blob/main/src/http/node/index.js it's only a few lines of code.

**Comment by mojavelinux** [comment]
While that's true, it uses quite a number of internal functions. That's why I pointed to my implementation, which is self-contained.

**Comment by mojavelinux** [comment]
I have documented how to write an http plugin for isomorphic-git that honors the HTTP proxy environment variables (http_proxy and https_proxy). This is the most minimal code I can come up with that will get it to work.

https://github.com/isomorphic-git/isomorphic-git/pull/1490

TLDR, here's the code for the plugin:

```js
'use strict'

const get = require('simple-get')
const { HttpProxyAgent, HttpsProxyAgent } = require('hpagent')

async function request ({ url, method, headers, body }) {
  body = await mergeBuffers(body)
  const proxy = url.startsWith('https:')
    ? { Agent: HttpsProxyAgent, url: process.env.https_proxy }
    : { Agent: HttpProxyAgent, url: process.env.http_proxy }
  const agent = proxy.url ? new proxy.Agent({ proxy: proxy.url }) : undefined
  return new Promise((resolve, reject) =>
    get({ url, method, agent, headers, body }, (err, res) => (err ? reject(err) : resolve(transformResponse(res))))
  )
}

async function mergeBuffers (data) {
  if (!Array.isArray(data)) return data
  if (data.length === 1 && data[0] instanceof Buffer) return data[0]
  const buffers = []
  let offset = 0
  let size = 0
  for await (const chunk of data) {
    buffers.push(chunk)
    size += chunk.byteLength
  }
  data = new Uint8Array(size)
  for (const buffer of buffers) {
    data.set(buffer, offset)
    offset += buffer.byteLength
  }
  return Buffer.from(data.buffer)
}

function transformResponse (res) {
  const { url, method, statusCode, statusMessage, headers } = res
  return { url, method, statusCode, statusMessage, headers, body: res }
}

module.exports = { request }
```

**Comment by mojavelinux** [comment]
See #1491 for a proposed change that would make this solution a lot simpler.
